utils/writeable/src/try_writeable.rs

```diff
@@ -0,0 +1,442 @@
+    ///     // see impl in TryWriteable docs
+    /// #    fn try_write_to_parts<S: PartsWrite + ?Sized>(
+    /// #        &self,
+    /// #        sink: &mut S,
+    /// #    ) -> Result<Result<(), Self::Error>, fmt::Error> {
+    /// #        sink.write_str("Hello, ")?;
+    /// #        // Use `impl TryWriteable for Result` to generate the error part:
+    /// #        let _ = self.name.ok_or("nobody").try_write_to_parts(sink)?;
+    /// #        sink.write_char('!')?;
+    /// #        // Return a doubly-wrapped Result.
+    /// #        // The outer Result is for fmt::Error, handled by the `?`s above.
+    /// #        // The inner Result is for our own Self::Error.
+    /// #        if self.name.is_some() {
+    /// #            Ok(Ok(()))
+    /// #        } else {
+    /// #            Ok(Err(MyWriteableError::MissingName))
+    /// #        }
+    /// #    }
+    /// #    fn try_writeable_length_hint(&self) -> (LengthHint, Option<Self::Error>) {
+    /// #        let (hint, e) = self.name.ok_or("nobody").try_writeable_length_hint();
+    /// #        (hint + 8, e.map(|_| MyWriteableError::MissingName))
+    /// #    }
+    /// }
+    ///
+    /// // Success case:
+    /// let writeable = HelloWorldWriteable { name: Some("Alice") };
+    /// let writeable_str = writeable.try_write_to_string().expect("name is Some");
+    ///
+    /// assert_eq!((Ordering::Equal, None), writeable.try_write_cmp_bytes(b"Hello, Alice!"));
+    ///
+    /// assert_eq!((Ordering::Greater, None), writeable.try_write_cmp_bytes(b"Alice!"));
+    /// assert_eq!(Ordering::Greater, (*writeable_str).cmp("Alice!"));
+    ///
+    /// assert_eq!((Ordering::Less, None), writeable.try_write_cmp_bytes(b"Hello, Bob!"));
+    /// assert_eq!(Ordering::Less, (*writeable_str).cmp("Hello, Bob!"));
+    ///
+    /// // Failure case:
+    /// let writeable = HelloWorldWriteable { name: None };
+    /// let mut writeable_str = String::new();
+    /// let _ = writeable.try_write_to(&mut writeable_str).expect("write to String is infallible");
+    ///
+    /// assert_eq!((Ordering::Equal, Some(MyWriteableError::MissingName)), writeable.try_write_cmp_bytes(b"Hello, nobody!"));
+    ///
+    /// assert_eq!((Ordering::Greater, Some(MyWriteableError::MissingName)), writeable.try_write_cmp_bytes(b"Hello, alice!"));
+    /// assert_eq!(Ordering::Greater, (*writeable_str).cmp("Hello, alice!"));
+    ///
+    /// assert_eq!((Ordering::Less, Some(MyWriteableError::MissingName)), writeable.try_write_cmp_bytes(b"Hello, zero!"));
+    /// assert_eq!(Ordering::Less, (*writeable_str).cmp("Hello, zero!"));
+    /// ```
+    fn try_write_cmp_bytes(&self, other: &[u8]) -> (Ordering, Option<Self::Error>) {
+        let mut wc = cmp::WriteComparator::new(other);
+        let result = match self.try_write_to(&mut wc) {
+            Ok(result) => result,
+            Err(core::fmt::Error) => {
+                debug_assert!(false, "WriteComparator infallible");
```

Comment:
same here

Reply:
I prefer being explicit about this being a `core::fmt::Error`, especially since there is a doubly nested result going on. I want to be clear that I'm not discarding the `Self::Error`.

Reply:
How about

```rust
        let result = self.try_write_to(&mut output).unwrap_or_else(|core::fmt::Error| Ok(()));
```

Reply:
I like it. Done